File: everything_parsing/07-chomsky-pumping/java/Pumping.java
/**
 * Module 07 — the pumping lemmas, run as an adversary.
 *
 * PUMPING LEMMA (regular): if L is regular, there is a length p such that every
 * s in L with |s| >= p splits as s = x y z with |y| >= 1, |xy| <= p, and
 * x y^k z is in L for ALL k >= 0.
 *
 * To prove { a^n b^n } is NOT regular we play the adversary: for ANY claimed p,
 * we hand back the string s = a^p b^p and show that EVERY legal (x, y, z) split
 * has some k with x y^k z not in L. The lemma cannot be satisfied for any p, so
 * L is not regular.  Same game, five parts, for the context-free lemma and
 * { a^n b^n c^n }.
 */
public final class Pumping {
    private Pumping() {}

    // ── the two target languages, as membership predicates ──

    static boolean inAnBn(String s) {
        int i = 0;
        while (i < s.length() && s.charAt(i) == 'a') i++;
        int a = i;
        while (i < s.length() && s.charAt(i) == 'b') i++;
        return i == s.length() && a == s.length() - a;
    }

    static boolean inAnBnCn(String s) {
        int i = 0;
        while (i < s.length() && s.charAt(i) == 'a') i++;
        int a = i;
        while (i < s.length() && s.charAt(i) == 'b') i++;
        int b = i - a;
        while (i < s.length() && s.charAt(i) == 'c') i++;
        int c = i - a - b;
        return i == s.length() && a == b && b == c;
    }

    // ── regular refutation for { a^n b^n } ──

    record RegularWitness(String s, String x, String y, String z, int k, String pumped,
                          int decompositionsTried, boolean allEscaped) {}

    static RegularWitness regularRefutation(int p) {
        String s = "a".repeat(p) + "b".repeat(p);
        int tried = 0;
        boolean allEscaped = true;
        String repX = "", repY = "", repZ = "", repPumped = "";
        int repK = 0;

        // every split s = x y z with |x y| = b <= p and |y| = b - a >= 1
        for (int b = 1; b <= p; b++) {
            for (int a = 0; a < b; a++) {
                tried++;
                String x = s.substring(0, a), y = s.substring(a, b), z = s.substring(b);
                int escapeK = firstEscape(x, y, z, Pumping::inAnBn);
                if (escapeK < 0) { allEscaped = false; }
                else if (repPumped.isEmpty()) {           // remember the first (a=0,b=1) as the representative
                    repX = x; repY = y; repZ = z; repK = escapeK;
                    repPumped = x + y.repeat(escapeK) + z;
                }
            }
        }
        return new RegularWitness(s, repX, repY, repZ, repK, repPumped, tried, allEscaped);
    }

    // ── context-free refutation for { a^n b^n c^n } ──

    record CflWitness(String s, String u, String v, String w, String x, String y,
                      int k, String pumped, int decompositionsTried, boolean allEscaped) {}

    static CflWitness cflRefutation(int p) {
        String s = "a".repeat(p) + "b".repeat(p) + "c".repeat(p);
        int n = s.length();
        int tried = 0;
        boolean allEscaped = true;
        String rU = "", rV = "", rW = "", rX = "", rY = "", rP = "";
        int rK = 0;

        // s = u v w x y, with v w x contiguous, |v w x| <= p, |v x| >= 1
        for (int i = 0; i <= n; i++) {                    // start of v
            for (int j = i; j <= Math.min(n, i + p); j++) {        // start of w
                for (int l = j; l <= Math.min(n, i + p); l++) {    // start of x
                    for (int m = l; m <= Math.min(n, i + p); m++) {// end of x
                        int vLen = j - i, xLen = m - l;
                        if (vLen + xLen < 1) continue;    // |v x| >= 1
                        if (m - i > p) continue;          // |v w x| <= p
                        tried++;
                        String u = s.substring(0, i), v = s.substring(i, j),
                               w = s.substring(j, l), x = s.substring(l, m), y = s.substring(m);
                        int esc = firstEscape5(u, v, w, x, y, Pumping::inAnBnCn);
                        if (esc < 0) allEscaped = false;
                        else if (rP.isEmpty()) {
                            rU = u; rV = v; rW = w; rX = x; rY = y; rK = esc;
                            rP = u + v.repeat(esc) + w + x.repeat(esc) + y;
                        }
                    }
                }
            }
        }
        return new CflWitness(s, rU, rV, rW, rX, rY, rK, rP, tried, allEscaped);
    }

    // ── helpers: smallest k in {0, 2, 3} with the pumped string outside L ──

    private interface InL { boolean test(String s); }

    private static int firstEscape(String x, String y, String z, InL inL) {
        for (int k : new int[]{2, 0, 3}) {
            if (!inL.test(x + y.repeat(k) + z)) return k;
        }
        return -1;
    }
    private static int firstEscape5(String u, String v, String w, String x, String y, InL inL) {
        for (int k : new int[]{2, 0, 3}) {
            if (!inL.test(u + v.repeat(k) + w + x.repeat(k) + y)) return k;
        }
        return -1;
    }
}

File: everything_parsing/07-chomsky-pumping/js/pumping.mjs
// Module 07 — the pumping lemmas, run as an adversary.
//
// Regular: if L is regular, some length p makes every s in L (|s| >= p) split
// as x y z with |y| >= 1, |xy| <= p and x y^k z in L for ALL k.
// We show { a^n b^n } cannot satisfy this for ANY p, and { a^n b^n c^n } cannot
// satisfy the 5-part context-free version for any p.

export function inAnBn(s) {
  let i = 0;
  while (i < s.length && s[i] === "a") i++;
  const a = i;
  while (i < s.length && s[i] === "b") i++;
  return i === s.length && a === s.length - a;
}

export function inAnBnCn(s) {
  let i = 0;
  while (i < s.length && s[i] === "a") i++;
  const a = i;
  while (i < s.length && s[i] === "b") i++;
  const b = i - a;
  while (i < s.length && s[i] === "c") i++;
  const c = i - a - b;
  return i === s.length && a === b && b === c;
}

const rep = (ch, n) => ch.repeat(n);

// ── regular refutation for { a^n b^n } ──

export function regularRefutation(p) {
  const s = rep("a", p) + rep("b", p);
  let tried = 0, allEscaped = true;
  let repX = "", repY = "", repZ = "", repPumped = "", repK = 0;

  for (let b = 1; b <= p; b++) {
    for (let a = 0; a < b; a++) {
      tried++;
      const x = s.slice(0, a), y = s.slice(a, b), z = s.slice(b);
      const k = firstEscape(kk => x + rep(y, kk) + z, inAnBn);
      if (k < 0) allEscaped = false;
      else if (repPumped === "") { repX = x; repY = y; repZ = z; repK = k; repPumped = x + rep(y, k) + z; }
    }
  }
  return { s, x: repX, y: repY, z: repZ, k: repK, pumped: repPumped, decompositionsTried: tried, allEscaped };
}

// ── context-free refutation for { a^n b^n c^n } ──

export function cflRefutation(p) {
  const s = rep("a", p) + rep("b", p) + rep("c", p);
  const n = s.length;
  let tried = 0, allEscaped = true;
  let rU = "", rV = "", rW = "", rX = "", rY = "", rP = "", rK = 0;

  for (let i = 0; i <= n; i++) {
    for (let j = i; j <= Math.min(n, i + p); j++) {
      for (let l = j; l <= Math.min(n, i + p); l++) {
        for (let m = l; m <= Math.min(n, i + p); m++) {
          const vLen = j - i, xLen = m - l;
          if (vLen + xLen < 1) continue;
          if (m - i > p) continue;
          tried++;
          const u = s.slice(0, i), v = s.slice(i, j), w = s.slice(j, l),
                x = s.slice(l, m), y = s.slice(m);
          const k = firstEscape(kk => u + rep(v, kk) + w + rep(x, kk) + y, inAnBnCn);
          if (k < 0) allEscaped = false;
          else if (rP === "") {
            rU = u; rV = v; rW = w; rX = x; rY = y; rK = k;
            rP = u + rep(v, k) + w + rep(x, k) + y;
          }
        }
      }
    }
  }
  return { s, u: rU, v: rV, w: rW, x: rX, y: rY, k: rK, pumped: rP, decompositionsTried: tried, allEscaped };
}

// smallest k in {2, 0, 3} that lands the pumped string outside L
function firstEscape(build, inL) {
  for (const k of [2, 0, 3]) if (!inL(build(k))) return k;
  return -1;
}
